Add strict parsing of aggregation ranges

cbuescher opened the pull request:
Currently we ignore unknown field names when parsing RangeAggregator.Range and GeoDistanceAggregationBuilder.Range from `range`, `date_range` or `geo_distance` aggregations. 
This can hide subtle errors, e.g. I ran into this yesterday by mistakenly trying to specify a "format" parameter but put it inside the range object, so it was ignored. To avoid this kind of mistakes we should make parsing strict here like we do with other queries, aggregations etc...

cbuescher committed:
Add strict parsing of aggregation ranges

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/GeoDistanceAggregationBuilder.java:
```diff
@@ -167,25 +172,38 @@ private static Range parseRange(XContentParser parser) throws IOException {
+                if (FROM_FIELD.match(currentFieldName) || TO_FIELD.match(currentFieldName)
+                        || KEY_FIELD.match(currentFieldName)) {
+                    // ignore null value
+                } else if (KEY_FIELD.match(currentFieldName)) {
```
is this condition not matched in the condition above? Maybe this should just be an else?

cbuescher replied:
Yes, thats a left-over. Good catch, this can go...

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregator.java:
```diff
@@ -151,7 +154,18 @@ public static Range fromXContent(XContentParser parser) throws IOException {
+                            || KEY_FIELD.match(currentFieldName)) {
+                        // ignore null value
+                    } else if (KEY_FIELD.match(currentFieldName)) {
+                        XContentParserUtils.throwUnknownField(currentFieldName, parser.getTokenLocation());
```
is this condition not matched in the condition above? Maybe this should just be an else?

colings86 reviewed (commented):
I like this change. I left a comment

cbuescher committed:
Adressing review comments

cbuescher committed:
Add test for parsing null values

cbuescher commented:
@colings86 thanks, I corrected the two problems you addressed and also added two more unit tests to check we parse `null` values in the ranges correctly. Turns out we currently only do that in a yaml test because in the base aggregation test the random builders ranges don't render "null".

colings86 reviewed (approved):
LGTM
